homeassistant/components/sma/__init__.py

```diff
@@ -1 +1,170 @@
+        # Add sensors from the custom config
+        # Supports deprecated yaml config from platform setup
+        sensor_def.add(
+            [
+                pysma.Sensor(
+                    o[CONF_KEY], n, o[CONF_UNIT], o[CONF_FACTOR], o.get(CONF_PATH)
+                )
+                for n, o in entry.data.get(CONF_CUSTOM).items()
+            ]
+        )
+
+        # When CONF_SENSORS is set, only enable sensors in config
+        # Supports deprecated yaml config from platform setup
+        config_sensors = entry.data.get(CONF_SENSORS)
+        if config_sensors:
+            # find and replace sensors removed from pysma
+            for sensor in config_sensors.copy():
+                if sensor in pysma.LEGACY_MAP:
+                    config_sensors.remove(sensor)
+                    config_sensors.append(pysma.LEGACY_MAP[sensor]["new_sensor"])
+
+            for s in sensor_def:
+                s.enabled = s.name in config_sensors
+
+        # Create list of all possible sensor names
+        possible_sensors = list(
+            set(config_sensors + [s.name for s in sensor_def] + list(pysma.LEGACY_MAP))
+        )
+
+        entity_registry = er.async_get(hass)
+
+        # Find entity_id using previous format of unique ID and change to new unique ID
+        for sensor in possible_sensors:
+            if sensor in sensor_def:
+                pysma_sensor = sensor_def[sensor]
+                original_key = pysma_sensor.key
+            elif sensor in pysma.LEGACY_MAP:
+                # If sensor was removed from pysma we will remap it to the new sensor
+                legacy_sensor = pysma.LEGACY_MAP[sensor]
+                pysma_sensor = sensor_def[legacy_sensor["new_sensor"]]
+                original_key = legacy_sensor["old_key"]
+            else:
+                _LOGGER.error("%s does not exist", sensor)
+                continue
+
+            entity_id = entity_registry.async_get_entity_id(
+                "sensor", "sma", f"sma-{original_key}-{sensor}"
+            )
+            if entity_id:
+                new_unique_id = (
+                    f"{entry.unique_id}-{pysma_sensor.key}_{pysma_sensor.key_idx}"
+                )
+                entity_registry.async_update_entity(
+                    entity_id, new_unique_id=new_unique_id
+                )
+
+    # Init the SMA interface
+    protocol = "https" if entry.data.get(CONF_SSL) else "http"
+    url = f"{protocol}://{entry.data.get(CONF_HOST)}"
+    verify_ssl = entry.data.get(CONF_VERIFY_SSL)
+    group = entry.data.get(CONF_GROUP)
+    password = entry.data.get(CONF_PASSWORD)
+
+    session = async_get_clientsession(hass, verify_ssl=verify_ssl)
+    sma = pysma.SMA(session, url, password, group)
+
+    # Ensure we logout on shutdown
+    async def async_close_session(event):
+        """Close the session."""
+        await sma.close_session()
+
+    hass.bus.async_listen(EVENT_HOMEASSISTANT_STOP, async_close_session)
```

Comment:
```suggestion
    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, async_close_session)
```

Reply:
Fixed in de5ecbe. Also removing the event listener now during unload